src/main/java/org/folio/circulation/services/AllowedServicePointsService.java

```diff
@@ -32,7 +60,92 @@ private CompletableFuture<Result<Map<RequestType, Set<String>>>> getAllowedServi
 
-    return ofAsync(new EnumMap<>(RequestType.class));
+    return userRepository.getUser(request.getRequesterId())
+      .thenCompose(r -> r.after(user -> lookupRequestPolicy(request, user)))
+      .thenCompose(r -> r.after(this::extractAllowedServicePoints));
+  }
+
+  private CompletableFuture<Result<RequestPolicy>> lookupRequestPolicy(
+    AllowedServicePointsRequest request, User user) {
+
+    log.debug("lookupRequestPolicy:: parameters: request={}, user={}", request, user);
+
+    if (user == null) {
+      log.error("lookupRequestPolicy:: user is null");
+      return completedFuture(failed(new RecordNotFoundFailure("User", request.getRequesterId())));
+    }
+
+    return itemRepository.fetchById(request.getItemId())
+      .thenCompose(r -> r.after(item -> requestPolicyRepository.lookupRequestPolicy(item, user)));
+  }
+
+  private CompletableFuture<Result<Map<RequestType, Set<String>>>> extractAllowedServicePoints(
+    RequestPolicy requestPolicy) {
+
+    log.debug("extractAllowedServicePoints:: parameters: requestPolicy={}", requestPolicy);
+
+    var allowedServicePointsInPolicy = requestPolicy.getAllowedServicePoints();
+    if (allowedServicePointsInPolicy.isEmpty()) {
+      log.info("extractAllowedServicePoints:: allowedServicePointsInPolicy is empty");
+
+      return fetchAllowedServicePoints(requestPolicy);
+    }
+
+    Set<String> allowedServicePointsIdsSet = allowedServicePointsInPolicy.values().stream()
+      .flatMap(Collection::stream)
+      .map(UUID::toString)
+      .collect(Collectors.toSet());
+
+    return servicePointRepository.filterIdsByServicePointsAndPickupLocationExistence(
+      allowedServicePointsIdsSet)
+        .thenApply(r -> r.next(filteredSpIds -> includeOnlyExistingServicePoints(
+          allowedServicePointsInPolicy, filteredSpIds, requestPolicy)));
+  }
+
+  private CompletableFuture<Result<Map<RequestType, Set<String>>>> fetchAllowedServicePoints(
+    RequestPolicy requestPolicy) {
+
+    log.debug("fetchAllowedServicePoints:: parameters: requestPolicy={}", requestPolicy);
+
+    List<RequestType> allowedTypes = Arrays.stream(RequestType.values())
+      .filter(requestPolicy::allowsType)
+      .toList();
+    if (allowedTypes.isEmpty()) {
+      log.info("fetchAllowedServicePoints:: allowedTypes is empty");
+      return ofAsync(new EnumMap<>(RequestType.class));
+    }
+    log.info("fetchAllowedServicePoints:: allowedTypes={}", allowedTypes);
+    return servicePointRepository.fetchPickupLocationServicePointsIds()
+      .thenCompose(r -> r.after(servicePointIds -> {
+        Map<RequestType, Set<String>> allowedServicePoints = new EnumMap<>(RequestType.class);
+        allowedTypes.forEach(requestType -> allowedServicePoints.put(requestType, servicePointIds));
+        log.info("fetchAllowedServicePoints:: result: {}", () -> asJson(allowedServicePoints));
+
+        return ofAsync(allowedServicePoints);
```

Comment:
- can we extract this block into a method to make code more readable?
- nothing async is happening, no need to wrap this code into a Future

Reply:
done